py/server/deephaven/experimental/data_index.py

```diff
@@ -0,0 +1,94 @@
+#
+# Copyright (c) 2016-2024 Deephaven Data Labs and Patent Pending
+#
+"""This module provides the ability to create, check, and retrieve DataIndex objects from Deephaven tables."""
+from typing import List, Optional
+
+import jpy
+
+from deephaven import DHError
+from deephaven._wrapper import JObjectWrapper
+from deephaven.jcompat import j_list_to_list
+from deephaven.table import Table
+
+_JDataIndexer = jpy.get_type("io.deephaven.engine.table.impl.indexer.DataIndexer")
+_JDataIndex = jpy.get_type("io.deephaven.engine.table.DataIndex")
+
+class DataIndex(JObjectWrapper):
+    """A DataIndex represents a Deephaven data index object which is built on one or more indexed (key) column(s) of
+    A Deephaven table."""
+    j_object_type = _JDataIndex
+
+    def __init__(self, j_data_index: jpy.JType):
+        self._j_data_index = j_data_index
+
+    @property
+    def j_object(self) -> jpy.JType:
+        return self._j_data_index
+
+    @property
+    def keys(self) -> List[str]:
+        """the key column names of the DataIndex.
+
+        Returns:
+            the key columns
+        """
+        return j_list_to_list(self._j_data_index.keyColumnNames())
+
+    @property
+    def table(self) -> Table:
+        """the backing table of the DataIndex.
+
+        Returns:
+            the backing table
+        """
```

Comment:
Documentation string is not well phrased to match the rest of our docs.  Changes may depend on what the method is called.